File: memory_model/sources/mem_model.4b.c
/*
* @@name:	mem_model.4b
* @@type:	C
* @@compilable:	yes
* @@linkable:	no
* @@expect:	success
* @@version:	omp_3.1
*/
#include <omp.h>

void flush_correct()
{
  int a, b;
  a = b = 0;
  #pragma omp parallel num_threads(2)
  {
    int myid = omp_get_thread_num();
    int tmp;

    if ( myid == 0 ) {          // thread 0
      #pragma omp atomic write
        b = 1;
      #pragma omp flush(a,b)    // flushes are ordered
      #pragma omp atomic read
        tmp = a;
    }
    else {                      // thread 1
      #pragma omp atomic write
        a = 1;
      #pragma omp flush(a,b)    // flushes are ordered
      #pragma omp atomic read
        tmp = b;
    }
    if ( tmp == 0 ) {           // access by single thread
      /* protected section */
    }
  }
}

File: memory_model/sources/mem_model.4b.f90
! @@name:	mem_model.4b
! @@type:	F-free
! @@operation:	compile
! @@expect:	success
! @@version:	omp_3.1
subroutine flush_correct
  use omp_lib
  implicit none
  integer a, b, tmp
  integer myid

  a = 0; b = 0
  !$omp parallel private(myid,tmp) num_threads(2)
    myid = omp_get_thread_num()

    if ( myid == 0 ) then     ! thread 0
      !$omp atomic write
        b = 1
      !$omp flush(a,b)        ! flushes are ordered
      !$omp atomic read
        tmp = a
    else                      ! thread 1
      !$omp atomic write
        a = 1
      !$omp flush(a,b)        ! flushes are ordered
      !$omp atomic read
        tmp = b
    endif
    if ( tmp == 0 ) then      ! access by single thread
      !! protected section
    endif
  !$omp end parallel
end subroutine
